### PR TITLE
Fix MO and nGraph to support the model context_rcnn_resnet101_snapshot_serengeti

### PR DESCRIPTION
**Pull request by vgavrilo**
Root cause analysis: 
When we convert the TF model [context_rcnn_resnet101_snapshot_serengeti](http://download.tensorflow.org/models/object_detection/context_rcnn_resnet101_snapshot_serengeti_2020_06_10.tar.gz)
using the command line
```sh
$ python3 path-to-mo/mo.py --input_model "context_rcnn_resnet101_snapshot_serengeti_2020_06_10/frozen_inference_graph.pb" --input "image_tensor[1 640 640 3],context_features[1 2000 2057],valid_context_size[1]"
```
we get an exception at the line 64 of `tools/mo/openvino/tools/mo/ops/select.py`, because the infer function of the MO operation `Select` doesn't take into an account that inputs for ports 1 and 2 can be scalars.

There is also the problem in the nGraph transformation `SharedShapeOf`: this transformation doesn't take into an account that collected nodes `ShapeOf` can have different values of the attribute `output_type`, that is some collected nodes have `output_type=element::i32`, and some ones have `output_type=element::i32`.

Solution:
Fix the infer function of the MO operation `Select` to support scalars in input ports `1` and `2`, and take into an account the attribute `output_type` value in the nGraph transformation `SharedShapeOf`.

Ticket: CVS-70255

Code:
* [x]  Comments
* [x]  Code style
* [x]  Transformation generates reshape-able IR - N/A
* [x]  Transformation preserves original framework node names - N/A

Validation:
* [x]  Unit tests
* [x]  Framework operation tests  - N/A
* [x]  Transformation tests - N/A
* [x]  Model Optimizer IR Reader check - N/A

Documentation:
* [x]  Supported frameworks operations list - N/A
* [x]  Guide on how to convert the **public** model - N/A
* [x]  User guide update - N/A

### CODE REVIEW

**Comment by achetver** [comment]
I think you should change `N/A` status to `updated` (or something like it) because you actually update them :)

**Review by achetver — APPROVED** [review]
MO part looks good.

**Comment by achetver** [comment]
Also, should we cover nGraph changes by tests?

**Review by pavel-esir — APPROVED** [review]

**Comment by GlebKazantaev** [comment]
@jane-intel please take a look.

**Review by jane-intel — CHANGES_REQUESTED** [review]
This is always true statement which was introduced in the original code. Please fix it
`root_ss->get_output_element_type(0) == root_ss->get_output_element_type(0)`

File: src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp [hunk 1]
```diff
@@ -38,10 +41,27 @@ bool ngraph::pass::SharedShapeOf::run_on_model(const std::shared_ptr<ngraph::Fun
...
         if (pair.second.size() < 2)
             continue;
-        const auto& root_ss = pair.second[0];
-        for (const auto& child_ss : pair.second)
-            if (root_ss->get_instance_id() != child_ss->get_instance_id() && root_ss->get_output_element_type(0) == root_ss->get_output_element_type(0))
-                graph_rewritten |= replace_output_update_name(child_ss->output(0), root_ss->output(0));
+
+        std::vector<std::shared_ptr<ngraph::Node>> nodes_for_different_types[2];
+        for (const auto& child : pair.second) {
+            const auto child_as_shapeof =  std::dynamic_pointer_cast<opset8::ShapeOf>(child);
```
Comment (jane-intel) [inline src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp:47]: what about opset1::ShapeOf?

**Comment by vgavrilo** [comment]
> This is always true statement which was introduced in the original code. Please fix it `root_ss->get_output_element_type(0) == root_ss->get_output_element_type(0)`

Done.

File: src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp (hunk 1 above)
Comment (vgavrilo) [reply src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp]: Fixed.

**Comment by vgavrilo** [comment]
> Also, should we cover nGraph changes by tests?

Written.

**Review by jane-intel — APPROVED** [review]

**Comment by vgavrilo** [comment]
@GlebKazantaev, @mvafin, please, see this PR (https://github.com/openvinotoolkit/openvino/pull/9255).

File: src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp [hunk 2]
```diff
@@ -38,10 +41,29 @@ bool ngraph::pass::SharedShapeOf::run_on_model(const std::shared_ptr<ngraph::Fun
     for (const auto& pair : source_to_shape_of) {
         if (pair.second.size() < 2)
             continue;
-        const auto& root_ss = pair.second[0];
-        for (const auto& child_ss : pair.second)
-            if (root_ss->get_instance_id() != child_ss->get_instance_id() && root_ss->get_output_element_type(0) == root_ss->get_output_element_type(0))
-                graph_rewritten |= replace_output_update_name(child_ss->output(0), root_ss->output(0));
+
+        std::vector<std::shared_ptr<ngraph::Node>> nodes_for_different_types[2];
```
Comment (GlebKazantaev) [inline src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp:45]: use `NodeVector`

File: src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp [hunk 3]
```diff
@@ -38,10 +41,29 @@ bool ngraph::pass::SharedShapeOf::run_on_model(const std::shared_ptr<ngraph::Fun
...
+            if (child_as_shapeof_v8) {
+                const auto& type_of_output = child_as_shapeof_v8->get_output_type();
+                size_t index = (type_of_output == element::i32) ? index_for_int32 : index_for_int64;
+                nodes_for_different_types[index].push_back(child);
+            } else {
+                const auto child_as_shapeof_v1 =  std::dynamic_pointer_cast<opset1::ShapeOf>(child);
+                if (child_as_shapeof_v1) {
+                    nodes_for_different_types[index_for_int64].push_back(child);
+                }
+            }
```
Comment (GlebKazantaev) [inline src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp:57]: Please use the same opsets for ShapeOf as mentioned above in the code.
```suggestion
            if (auto shapeof = std::dynamic_pointer_cast<opset3::ShapeOf>(child)) {
                const auto& type_of_output = shapeof ->get_output_type();
                size_t index = (type_of_output == element::i32) ? index_for_int32 : index_for_int64;
                nodes_for_different_types[index].push_back(child);
            } else if (auto shapeof = std::dynamic_pointer_cast<opset1::ShapeOf>(child)) {
                 nodes_for_different_types[index_for_int64].push_back(child);
            }
```

**Review by GlebKazantaev — APPROVED** [review]

File: src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp (hunk 3 above)
Comment (GlebKazantaev) [reply src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp:57]: BTW, can we use just `child->output_type()` to determine the index without switches?

File: src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp (hunk 2 above)
Comment (vgavrilo) [reply src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp]: Done.

File: src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp (hunk 3 above)
Comment (vgavrilo) [reply src/common/transformations/src/transformations/common_optimizations/simplify_shape_of_sub_graph.cpp]: Done.

**Pull request merged**